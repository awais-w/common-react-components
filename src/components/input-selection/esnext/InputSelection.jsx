import React from 'react';
import PropTypes from 'prop-types';
import bem from '@argos/utils/esnext/bem';
import './InputSelection.scss';

const propTypes = {
  checked: PropTypes.bool.isRequired,
  children: PropTypes.any,
  checkboxIconType: PropTypes.oneOf(['tick', 'cross', 'line']),
  id: PropTypes.string.isRequired,
  inputType: PropTypes.oneOf(['checkbox', 'radio']).isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
};

const defaultProps = {
  checkboxIconType: 'tick',
};

const InputSelection = (props) => {
  const { checked, children, checkboxIconType, id, inputType, name, onChange, value } = props;
  const cn = bem({ prefix: 'cmc-', block: 'input-selection' });
  return (
    <div className={cn()}>
      <input
        checked={checked}
        className={cn('input', inputType === 'checkbox' ? checkboxIconType : '')}
        id={id}
        name={name}
        onChange={onChange}
        type={inputType}
        value={value}
      />
      <label htmlFor={id} className={cn('label')}>
        <div className={cn('text-container')}>{children}</div>
      </label>
    </div>
  );
};

InputSelection.defaultProps = defaultProps;
InputSelection.propTypes = propTypes;
export default InputSelection;
