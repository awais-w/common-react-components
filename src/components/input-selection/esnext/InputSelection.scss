@import '@cmc/common-styles/src/scss/_variables.scss';

.cmc-input-selection {
  position: relative;
  display: block;

  label {
    @include font-standard;
    color: $color-grey-charcoal;
    cursor: pointer;
    margin: 0;
    padding: 0;
    display: flex;
    align-items: center;

    :focus {
      outline: inherit;
      outline-offset: $half;
    }
  }

  .cmc-input-selection__text-container {
    padding-left: $base;
  }

  input[type='checkbox'],
  input[type='radio'] {
    margin-right: $base;
    opacity: 0;
    position: absolute;
    align-self: center;

    & + label:before {
      content: '';
      @include prefix(transition, all 0.3s ease, webkit moz ms o);
      font-family: 'Material Icons';
      color: $color-white;
      display: block;
      width: $input-height-regular * 0.5;
      height: $input-height-regular * 0.5;
      background-color: $color-white;
      border: 1px solid $color-grey-dark;
      display: flex;
      align-items: center;
      justify-content: center;
      border-radius: $small;
      box-sizing: border-box;
      @include touch-device-bp1024 {
        width: $large;
        height: $large;
      }
    }

    &:active + label:before {
      box-shadow: 0;
    }

    &:checked.cmc-input-selection__input--tick + label:before {
      content: 'check';
      background-color: $primary-color;
      border-color: $primary-color;
    }

    &:checked.cmc-input-selection__input--cross + label:before {
      content: 'close';
      background-color: $color-rust;
      border-color: $color-rust;
    }

    &:checked.cmc-input-selection__input--line + label:before {
      content: 'remove';
      background-color: $color-sainsburys;
      border-color: $color-sainsburys;
    }

    :focus + label:before {
      border-color: $color-grey-dark;
    }
  }

  input[type='radio'] {
    & + label:before {
      border-radius: 50%;
      content: '';
    }
  }

  input[type='radio']:checked {
    & + label:before {
      border: $half + 2 solid $primary-color;
      @include touch-device-bp1024 {
        border-width: $base;
      }
    }
  }

  input input:focus + label {
    transition: none 0s ease 0s;
  }

  fieldset[disabled] label,
  fieldset[disabled] input,
  input.disabled,
  input[disabled] {
    cursor: not-allowed;
  }
}
