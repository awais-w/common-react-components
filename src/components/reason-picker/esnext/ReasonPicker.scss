@import '@cmc/common-styles/src/scss/_variables.scss';
.cmc-reason-picker {
  margin-bottom: $double;
  &__line {
    display: grid;
    grid-template-columns: 3fr 2fr;
    margin-bottom: $base;
    grid-column-gap: $base;
    align-items: center;

    select {
      margin-bottom: 0;
      font-size: $font-size-body;
      align-self: center;
      height: $input-height-regular;
    }

    textarea {
      margin: 0;
      resize: vertical;
      height: $input-height-regular;
    }

    strong {
      font-weight: bold;
    }

    .cmc-input-selection {
      &__text-container {
        > span:not(:last-child) {
          margin-right: $base;
        }
      }
    }
  }
}
