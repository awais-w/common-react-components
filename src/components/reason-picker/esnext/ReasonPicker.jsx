import React from 'react';
import PropTypes from 'prop-types';
import bem from '@argos/utils/esnext/bem';
import InputSelection from '@cmc/input-selection/esnext/InputSelection';
import uuid from 'uuid';
import './ReasonPicker.scss';

const cn = bem({ prefix: 'cmc', block: '-reason-picker' });

const ReasonPicker = ({
  listItems,
  reasonItems,
  reasonPlaceholderTxt,
  onListItemClick,
  onReasonChange,
  checkboxIconType,
  inputType,
  reasonType,
}) => {
  const dropDown = (listItemId, reasonSelectedId) => {
    const selectedReason = reasonItems.find((reason) =>
      reasonSelectedId ? reason.id === reasonSelectedId : reason.selected,
    );
    return (
      <select
        className='form-control'
        defaultValue={selectedReason && selectedReason.id}
        onChange={(event) => onReasonChange(event, listItemId)}>
        {reasonItems.map((reason) => {
          const { id, description } = reason;
          return (
            <option value={id} key={`option-${id}`}>
              {description}
            </option>
          );
        })}
      </select>
    );
  };

  const uniqueId = uuid.v4();

  return (
    <div className={cn()}>
      <ul>
        {listItems.map((item) => {
          const { id, content, reasonSelectedId, selected } = item;
          return (
            <li className={cn('line')} key={id}>
              <InputSelection
                checkboxIconType={checkboxIconType}
                id={`input-${uuid.v4()}`}
                inputType={inputType}
                value={id}
                name={inputType === 'radio' ? `radio-selection-${uniqueId}` : `checkbox-${uuid.v4()}`}
                onChange={(event) => onListItemClick(event, id, inputType)}
                checked={!!selected}>
                {content.map((token, index) => (
                  <span key={`${id}-${index}`}>
                    <strong>{token.key}</strong>
                    {token.value}
                  </span>
                ))}
              </InputSelection>
              {!!selected && (
                <React.Fragment key={`select-${id}`}>
                  {reasonType === 'list' ? (
                    dropDown(id, reasonSelectedId)
                  ) : (
                    <span className={cn('textarea')}>
                      <textarea
                        className='form-control'
                        name={`textarea-${uuid.v4()}`}
                        placeholder={reasonPlaceholderTxt}
                        id={`textarea-${id}`}
                        onChange={(event) => onReasonChange(event, id)}
                      />
                    </span>
                  )}
                </React.Fragment>
              )}
            </li>
          );
        })}
      </ul>
    </div>
  );
};

ReasonPicker.propTypes = {
  listItems: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      content: PropTypes.arrayOf(
        PropTypes.shape({
          key: PropTypes.string.isRequired,
          value: PropTypes.string.isRequired,
        }),
      ),
      selected: PropTypes.bool,
      reasonSelectedId: PropTypes.string,
    }),
  ),
  reasonItems: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      selected: PropTypes.bool,
    }),
  ),
  onListItemClick: PropTypes.func,
  onReasonChange: PropTypes.func,
  reasonPlaceholderTxt: PropTypes.string,
  checkboxIconType: PropTypes.oneOf(['tick', 'cross']),
  inputType: PropTypes.oneOf(['checkbox', 'radio']).isRequired,
  reasonType: PropTypes.oneOf(['list', 'text']).isRequired,
};

ReasonPicker.defaultProps = {
  inputType: 'radio',
  reasonType: 'list',
  listItems: [],
  reasonItems: [],
  isItemChecked: () => {},
  onListItemClick: () => {},
  onReasonChange: () => {},
};

export default ReasonPicker;
