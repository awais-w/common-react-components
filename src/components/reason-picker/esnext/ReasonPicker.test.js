import React from 'react';
import { shallow } from 'enzyme';
import uuid from 'uuid';
import ReasonPicker from './ReasonPicker';
import InputSelection from '../../input-selection/esnext/InputSelection';
import { listItems, reasonItems } from './ReasonPicker.data.js';

uuid.v4 = jest.fn(() => '00000000-0000-0000-0000-000000000000');

const defaultProps = {
  inputType: 'radio',
  listItems,
  reasonItems,
  reasonPlaceholderTxt: 'Please state reason',
  onReasonChange: jest.fn(),
  onListItemClick: () => {},
  reasonType: 'list',
};

const setup = (props) => {
  const component = shallow(<ReasonPicker {...props} />);

  return {
    wrapper: component,
    getInputSelect: () => component.find(InputSelection),
    getLine: () => component.find('.cmc-reason-picker__line'),
    getSelection: () => component.find('.cmc-reason-picker__selection'),
    getSelect: () => component.find('select.form-control'),
    getTextArea: () => component.find('textarea.form-control'),
    getTextAreaPlaceholder: () => component.find('textarea.form-control').prop('placeholder'),
  };
};

describe('<ReasonPicker />', () => {
  const component = setup(defaultProps);

  it('should match default snapshot', () => {
    expect(component.wrapper).toMatchSnapshot();
  });

  describe('rendering', () => {
    it('renders the correct number of lines', () => {
      expect(component.getLine()).toHaveLength(3);
    });

    it('renders the correct number of drop downs', () => {
      expect(component.getSelect()).toHaveLength(1);
    });

    it('renders the correct number of textareas', () => {
      component.wrapper.setProps({ reasonType: 'text' });
      expect(component.getTextArea()).toHaveLength(1);
    });

    it('renders placeholder of textarea', () => {
      expect(component.getTextAreaPlaceholder()).toBe('Please state reason');
    });

    it('renders the InputSelect component', () => {
      expect(component.getInputSelect()).toHaveLength(3);
    });
  });
});

describe('reasonChange functionality', () => {
  const component = setup(defaultProps);
  it('onReasonChange function prop is called when the select input is changed', () => {
    component.getSelect().simulate('change');
    expect(defaultProps.onReasonChange).toBeCalled();
  });
  it('onReasonChange function prop is called when the textArea input is changed', () => {
    component.wrapper.setProps({ reasonType: 'text' });
    component.getTextArea().simulate('change');
    expect(defaultProps.onReasonChange).toBeCalled();
  });
});
