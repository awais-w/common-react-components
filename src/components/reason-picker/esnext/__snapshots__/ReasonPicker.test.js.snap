// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReasonPicker /> should match default snapshot 1`] = `
<div
  className="cmc-reason-picker"
>
  <ul>
    <li
      className="cmc-reason-picker__line"
      key="1a"
    >
      <InputSelection
        checkboxIconType="tick"
        checked={false}
        id="input-00000000-0000-0000-0000-000000000000"
        inputType="radio"
        name="radio-selection-00000000-0000-0000-0000-000000000000"
        onChange={[Function]}
        value="1a"
      >
        <span
          key="1a-0"
        >
          <strong>
            Order:
          </strong>
          1234
        </span>
        <span
          key="1a-1"
        >
          <strong>
            Description:
          </strong>
          A very bad thing
        </span>
      </InputSelection>
    </li>
    <li
      className="cmc-reason-picker__line"
      key="2a"
    >
      <InputSelection
        checkboxIconType="tick"
        checked={true}
        id="input-00000000-0000-0000-0000-000000000000"
        inputType="radio"
        name="radio-selection-00000000-0000-0000-0000-000000000000"
        onChange={[Function]}
        value="2a"
      >
        <span
          key="2a-0"
        >
          <strong>
            Order:
          </strong>
          23456
        </span>
        <span
          key="2a-1"
        >
          <strong>
            Description:
          </strong>
          A very nice thing
        </span>
      </InputSelection>
      <select
        className="form-control"
        defaultValue="r1a"
        onChange={[Function]}
      >
        <option
          key="option-0"
          value="0"
        >
          Please select
        </option>
        <option
          key="option-r1a"
          value="r1a"
        >
          I just don't want it
        </option>
        <option
          key="option-r2a"
          value="r2a"
        >
          I already have one
        </option>
        <option
          key="option-r3a"
          value="r3a"
        >
          It is not what I expected
        </option>
      </select>
    </li>
    <li
      className="cmc-reason-picker__line"
      key="3a"
    >
      <InputSelection
        checkboxIconType="tick"
        checked={false}
        id="input-00000000-0000-0000-0000-000000000000"
        inputType="radio"
        name="radio-selection-00000000-0000-0000-0000-000000000000"
        onChange={[Function]}
        value="3a"
      >
        <span
          key="3a-0"
        >
          <strong>
            Order:
          </strong>
          34567
        </span>
        <span
          key="3a-1"
        >
          <strong>
            Description:
          </strong>
          It is ok I guess
        </span>
      </InputSelection>
    </li>
  </ul>
</div>
`;
